sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/ShareFileAsyncClient.java

```diff
@@ -1309,27 +1331,228 @@ private void channelCleanUp(AsynchronousFileChannel channel) {
      * status code.
      * @throws ShareStorageException If you attempt to upload a range that is larger than 4 MB, the service returns
      * status code 413 (Request Entity Too Large)
+     *
+     * @deprecated Use {@link ShareFileAsyncClient#uploadRangeWithResponse(ShareFileUploadRangeOptions)} instead. Or
+     * consider {@link ShareFileAsyncClient#uploadWithResponse(ShareFileUploadOptions)} for a more capable upload
+     * method.
      */
+    @Deprecated
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<ShareFileUploadInfo>> uploadWithResponse(Flux<ByteBuffer> data, long length, Long offset,
         ShareRequestConditions requestConditions) {
         try {
-            return withContext(context -> uploadWithResponse(data, length, offset, requestConditions,
-                context));
+            return withContext(context -> uploadRangeWithResponse(new ShareFileUploadRangeOptions(data, length)
+                .setOffset(offset).setRequestConditions(requestConditions), context));
         } catch (RuntimeException ex) {
             return monoError(logger, ex);
         }
     }
 
-    Mono<Response<ShareFileUploadInfo>> uploadWithResponse(Flux<ByteBuffer> data, long length, Long offset,
-        ShareRequestConditions requestConditions, Context context) {
-        requestConditions = requestConditions == null ? new ShareRequestConditions() : requestConditions;
-        long rangeOffset = (offset == null) ? 0L : offset;
-        ShareFileRange range = new ShareFileRange(rangeOffset, rangeOffset + length - 1);
+    /**
+     * Buffers a range of bytes and uploads sub-ranges in parallel to a file in storage file service. Upload operations
+     * perform an in-place write on the specified file.
+     *
+     * <p><strong>Code Samples</strong></p>
+     *
+     * <p>Upload data "default" to the file in Storage File Service. </p>
+     *
+     * {@codesnippet com.azure.storage.file.share.ShareFileAsyncClient.upload#Flux-ParallelTransferOptions}
+     *
+     * <p>For more information, see the
+     * <a href="https://docs.microsoft.com/rest/api/storageservices/put-range">Azure Docs</a>.</p>
+     *
+     * @param data The data which will upload to the storage file.
+     * @param transferOptions {@link ParallelTransferOptions} to use to upload data.
+     * @return The {@link ShareFileUploadInfo file upload info}
+     */
+    public Mono<ShareFileUploadInfo> upload(Flux<ByteBuffer> data, ParallelTransferOptions transferOptions) {
```

Comment:
We should add the boolean overwrite overloads as well for async and sync.

Reply:
I think that's a blobs only concept. I can't find an overload on this client that accepts an override parameter. File writes just override the range, and you have to successfully create a file before you can put a range.

Reply:
Hmm, yeah I think the feature just never made its way to file shares for some reason. I personally think we should still do it to save our customers (since we do it in blobs and datalake already). @alzimmermsft, @srnagar, @kasobol-msft 
any thoughts?

Reply:
It's a thing in both blobs and datalake. It isn't in queues (but that's reasonable). I think we should have it here.